### PR TITLE
Config system

### PR DESCRIPTION
**Pull request by zkxs**
[Original discussion](https://github.com/zkxs/NeosModLoader/discussions/4)  
[Usage example](https://github.com/zkxs/NeosModConfigurationExample)  
[Documentation preview](https://github.com/zkxs/NeosModLoader/blob/config-system/doc/config.md) (link will change once merged)

This PR adds a configuration system to the modloader. Configs are stored as json files in a `nml_config` directory. These json files are not intended to be directly manipulated by end users or mods. Instead, an API has been added to NeosModLoader itself that allows defining, enumerating, getting, and setting configs.

Once the community is done eyeballing this PR I'll release it in NeosModLoader 1.5.0

### CODE REVIEW

**Comment by Lexevolution** [comment]
I'm guessing a mod would be able to load its config at any arbitrary point in time? (more specifically, the ability to load from storage, not already in memory, to check for modified values)

And for writing values to the config, it seems like the value type has to be manually defined by the mod creator? 
Is there a feature built in that converts complex Neos data types to ones accepted by the JSON config? Or does that have to be handled by the mod creator in their mod?

File: NeosModLoader/NeosModLoader.csproj [hunk 1]
```diff
@@ -40,18 +40,27 @@
     <Reference Include="FrooxEngine">
       <HintPath>C:\Program Files (x86)\Steam\steamapps\common\NeosVR\Neos_Data\Managed\FrooxEngine.dll</HintPath>
     </Reference>
+    <Reference Include="Newtonsoft.Json">
+      <HintPath>C:\Program Files (x86)\Steam\steamapps\common\NeosVR\Neos_Data\Managed\Newtonsoft.Json.dll</HintPath>
```
Comment (Lexevolution) [inline NeosModLoader/NeosModLoader.csproj:44]: Would this hardcoded location be an issue for people who have their Neos install somewhere else? I'm not 100% sure how `<HintPath>` works with external libraries.

Comment (EIA485) [reply NeosModLoader/NeosModLoader.csproj:44]: nope, this is just the path used for compiling.

**Comment by zkxs** [comment]
@Lexevolution 
> I'm guessing a mod would be able to load its config at any arbitrary point in time? (more specifically, the ability to load from storage, not already in memory, to check for modified values)

The configs are read from disk into memory once when the modloader initializes. From that point on you cannot reload from disk, only save to disk. All mods who are using a config are using the same instance, so they'll be able to see each other's changes. Also, as of 8c9eeb4  and 701b88f there is now an event system mods can use to subscribe to configuration value change events.

> And for writing values to the config, it seems like the value type has to be manually defined by the mod creator? Is there a feature built in that converts complex Neos data types to ones accepted by the JSON config? Or does that have to be handled by the mod creator in their mod?

In theory Json.NET can serialize complex types. In practice some types are going to have issues. For example, you cannot represent a circularly linked list in Json (JSON is a tree, and circular references require a graph). I also suspect using complex types will be a thorn in the side of any configuration manager mods people make, even if Json.NET is able to serialize them.

**Pull request merged**